examples/kuadrant/reconciler.go

```diff
@@ -0,0 +1,260 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"sort"
+	"strings"
+
+	egv1alpha1 "github.com/envoyproxy/gateway/api/v1alpha1"
+	"github.com/google/go-cmp/cmp"
+	"github.com/samber/lo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/utils/ptr"
+	gwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+	gwapiv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+
+	"github.com/kuadrant/policy-machinery/controller"
+	"github.com/kuadrant/policy-machinery/machinery"
+
+	kuadrantapis "github.com/kuadrant/policy-machinery/examples/kuadrant/apis"
+	kuadrantv1alpha2 "github.com/kuadrant/policy-machinery/examples/kuadrant/apis/v1alpha2"
+	kuadrantv1beta3 "github.com/kuadrant/policy-machinery/examples/kuadrant/apis/v1beta3"
+)
+
+const topologyFile = "topology.dot"
+
+type GatewayProvider interface {
+	ReconcileGateway(topology *machinery.Topology, gateway machinery.Targetable, effectivePolicies map[string][]machinery.Policy)
+}
+
+type Reconciler struct {
+	GatewayProvider GatewayProvider
+}
+
+func (r *Reconciler) Reconcile(eventType controller.EventType, oldObj, newObj controller.RuntimeObject, topology *machinery.Topology) {
+	// print the event
+	obj := oldObj
+	if obj == nil {
+		obj = newObj
+	}
+	log.Printf("%s %sd: %s/%s\n", obj.GetObjectKind().GroupVersionKind().Kind, eventType.String(), obj.GetNamespace(), obj.GetName())
+	if eventType == controller.UpdateEvent {
+		log.Println(cmp.Diff(oldObj, newObj))
+	}
+
+	// update the topology file
+	saveTopologyToFile(topology)
+
+	targetables := topology.Targetables()
+
+	// reconcile policies
+	gateways := targetables.Items(func(o machinery.Object) bool {
+		_, ok := o.(*machinery.Gateway)
+		return ok
+	})
+
+	listeners := targetables.Items(func(o machinery.Object) bool {
+		_, ok := o.(*machinery.Listener)
+		return ok
+	})
+
+	httpRouteRules := targetables.Items(func(o machinery.Object) bool {
+		_, ok := o.(*machinery.HTTPRouteRule)
+		return ok
+	})
+
+	effectivePolicies := map[string][]machinery.Policy{}
+
+	for _, gateway := range gateways {
+		// reconcile Gateway -> Listener policies
+		for _, listener := range listeners {
+			paths := targetables.Paths(gateway, listener)
+			for i := range paths {
+				if p := effectivePolicyForPath[*kuadrantv1alpha2.DNSPolicy](paths[i]); p != nil {
+					effectivePolicies["dns"] = append(effectivePolicies["dns"], *p)
+				}
+				if p := effectivePolicyForPath[*kuadrantv1alpha2.TLSPolicy](paths[i]); p != nil {
+					effectivePolicies["tls"] = append(effectivePolicies["tls"], *p)
+				}
+			}
+		}
```

Comment:
Found an edge case that probably is not intended which suggests an issue around here I think 🤔 

If following the `envoy-gateway.md` we create the `other-gateway` with the unmanged GatewayClass  first instead and attach the gateway AuthPolicy to this gateway. No security policy is created which is expected.

Now if we create the managed envoy gateway class and gateway, we see a security policy created for this gateway which is not expected I think since there's no AuthPolicy affecting this managed gateway. Seems like the auth effective policy from the unmanged gateway is calculated and passed to this gateway because we loop through **all** listeners instead of listeners associated only to the looped gateway 🤔 

<img width="609" alt="image" src="https://github.com/user-attachments/assets/a4119bf2-d446-4ef2-b51f-7cdc1a914edf">


Replication steps:
1. Create Gateway that is not managed by Envoy Gateway:

```sh
kubectl apply -f -<<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: GatewayClass
metadata:
  name: istio
spec:
  controllerName: istio.io/gateway-controller
EOF
```

```sh
kubectl apply -f -<<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: other-gateway
spec:
  gatewayClassName: istio
  listeners:
  - name: http
    protocol: HTTP
    port: 80
    allowedRoutes:
      namespaces:
        from: Same
EOF
```

2. Create a HTTPRoute:

```sh
kubectl apply -f -<<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: my-app
spec:
  parentRefs:
  - name: other-gateway
  hostnames:
  - example.com
  rules:
  - matches:
    - method: POST
    - method: GET
    backendRefs:
    - name: my-app
      port: 80
EOF
```

3. Create a Gateway-wide AuthPolicy allowing access to the services between 8am to 5pm only by default:

```sh
kubectl apply -f - <<EOF
apiVersion: kuadrant.io/v1beta3
kind: AuthPolicy
metadata:
  name: business-hours
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: other-gateway
  overrides:
    rules:
      authorization:
        "from8am-to-5pm":
          opa:
            rego: |
              allow { [h, _, _] := time.clock(time.now_ns()); h >= 8; h <= 17 }
    strategy: merge
EOF
```

4. Create a Gateway managed by the Envoy Gateway gateway controller:

```sh
kubectl apply -f -<<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: GatewayClass
metadata:
  name: eg
spec:
  controllerName: gateway.envoyproxy.io/gatewayclass-controller
EOF
```

```sh
kubectl apply -f -<<EOF
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  name: prod-web
spec:
  gatewayClassName: eg
  listeners:
  - name: http
    protocol: HTTP
    port: 80
    allowedRoutes:
      namespaces:
        from: Same
EOF
```

Reply:
Yep. I've spotted that myself a couple hours ago.

I already have a fix which I will soon push including as well some refactoring (there's need to pass the effective policies to `ReconcileGateway`) and an extension of the example supporting Istio-managed gateways too.

Thanks for digging into it, @KevFan!

Reply:
Oh, nice! 👍

Reply:
Done